contracts/libs/Types.sol

```diff
@@ -157,19 +157,39 @@ library Types {
         uint256 fee;
     }
 
-    // UserAccount contains the actual data stored in the leaf of balance tree
-    struct UserAccount {
-        // ID is the path to the pubkey in the PDA tree
-        uint256 ID;
+    /**
+    @param pubkeyIndex path to the pubkey in the PDA tree
+     */
+    struct UserState {
+        uint256 pubkeyIndex;
         uint256 tokenType;
         uint256 balance;
         uint256 nonce;
     }
 
-    struct AccountMerkleProof {
-        UserAccount account;
-        uint256 pathToAccount; // This field is kept for backward competibility, don't use it.
-        bytes32[] siblings;
+    function encode(UserState memory state)
```

Comment:
We just use `state.encode()` everywhere. We use `keccak256(state.encode())` but not `state.hash()` so we don't either use 2 function jumps or rewrite abi.encode twice.